plugins/catalog-backend-module-azure-blobstorage/src/provider/AzureBlobStorageEntityProvider.ts

```diff
@@ -0,0 +1,192 @@
+        options.logger,
+        taskRunner,
+      );
+    });
+  }
+
+  private constructor(
+    private readonly config: AzureBlobStorageConfig,
+    integration: AzureBlobStorageIntegration,
+    logger: Logger,
+    taskRunner: TaskRunner,
+  ) {
+    this.integration = integration;
+    this.logger = logger.child({
+      target: this.getProviderName(),
+    });
+    this.storageClient = new BlobServiceClient(
+      `https://${integration?.config?.host}${integration?.config?.secretAccessKey}`,
+    );
+
+    this.scheduleFn = this.createScheduleFn(taskRunner);
+  }
+
+  private createScheduleFn(taskRunner: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return taskRunner.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.logger.child({
+            class: AzureBlobStorageEntityProvider.prototype.constructor.name,
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(`${this.getProviderName()} refresh failed`, error);
+          }
+        },
+      });
+    };
+  }
+
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    await this.scheduleFn();
+  }
+
+  getProviderName(): string {
+    return `azureBlobStorage-provider:${this.config.accountName}`;
+  }
+
+  async refresh(logger: Logger) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    logger.info('Discovering Azure BlobStorage objects');
+
+    const keys = await this.listAllObjectKeys();
+    logger.info(`Discovered ${keys.length} Azure BlobStorage objects`);
+
+    const locations = keys.map(key => this.createLocationSpec(key));
+
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: locations.map(location => {
+        return {
+          locationKey: this.getProviderName(),
+          entity: locationSpecToLocationEntity({ location }),
+        };
+      }),
+    });
+
+    logger.info(
+      `Committed ${locations.length} Locations for Azure BlobStorage objects`,
+    );
+  }
+
+  private async listAllObjectKeys(): Promise<string[]> {
+    const keys: string[] = [];
+
+    const containerClient: ContainerClient =
+      this.storageClient.getContainerClient(this.config.containerName);
+    const blobs = containerClient.listBlobsFlat({ prefix: this.config.prefix });
+    for await (const blob of blobs) {
+      keys.push(`${this.config.containerName}/${blob.name}`);
+    }
+
+    return keys;
+  }
+
+  private createLocationSpec(key: string): LocationSpec {
+    return {
+      type: 'url',
+      target: this.createObjectUrl(key),
+      presence: 'required',
+    };
+  }
+
+  private createObjectUrl(key: string): string {
+    return encodeURI(`https://${this.integration?.config?.host}/${key}`);
```

Comment:
hm why the encodeURI? It's a bit of an oddball method that I don't generally trust, and is it necessary here?